go/tools/builders/link.go

```diff
@@ -51,16 +50,47 @@ func link(args []string) error {
 	buildmode := flags.String("buildmode", "", "Build mode used.")
 	flags.Var(&xdefs, "X", "A string variable to replace in the linked binary (repeated).")
 	flags.Var(&stamps, "stamp", "The name of a file with stamping values.")
-	conflictErrMsg := flags.String("conflict_err", "", "Error message about conflicts to report if there's a link error.")
 	if err := flags.Parse(builderArgs); err != nil {
 		return err
 	}
 	if err := goenv.checkFlagsAndSetGoroot(); err != nil {
 		return err
 	}
 
-	if *conflictErrMsg != "" {
-		return errors.New(*conflictErrMsg)
+	var dedupedArchives []archive
+	seenArchives := make(map[string]archive)
+	for _, arc := range archives {
+		seen, ok := seenArchives[arc.packagePath]
+		if !ok {
+			dedupedArchives = append(dedupedArchives, arc)
+			seenArchives[arc.packagePath] = arc
+			continue
+		}
+
+		if arc.packagePath == "github.com/bazelbuild/rules_go/go/tools/coverdata" {
+			// We can end up with this archive twice if the test transitively depends on coverdata.
+			// See https://github.com/bazelbuild/rules_go/issues/3017 and
+			// https://github.com/bazelbuild/rules_go/pull/3032
+			// Note that `arc.file` != `seen.file`:
+			//   _main/bazel-out/darwin_arm64-fastbuild-ST-92a08095f520/bin/external/io_bazel_rules_go/go/tools/coverdata/coverdata.a
+			//   _main/bazel-out/darwin_arm64-fastbuild/bin/external/io_bazel_rules_go/go/tools/coverdata/coverdata.a
+			// Perhaps path mapping can help here?
```

Comment:
It can help avoid the overhead, but not the potential correctness problem in case any of these is affected by different Go settings. Could you diff the configs of the two files? Maybe we should add them in a different place that isn't subject to this duplication.

Reply:
ok I pushed an extra commit fixing it by transitioning `coverdata` to non_go_tool cfg, wdyt?